### PR TITLE
Improve the `synapse.api.auth.Auth` mock used in unit tests.

### PR DESCRIPTION
**Pull request by sandhose**
This enhances the mocked `synapse.api.auth.Auth` module, mainly by returning a `Requester` on `get_user_by_access_token` instead of a `dict`. It also uses named parameters on `create_requester`, and does not specify the default values anymore.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: tests/unittest.py [hunk 1]
```diff
@@ -300,46 +300,39 @@ def setUp(self) -> None:
...
-                        False,
-                        False,
-                        None,
-                    )
+                    return await get_user_by_access_token(token)
 
                 # Type ignore: mypy doesn't like us assigning to methods.
                 self.hs.get_auth().get_user_by_req = get_user_by_req  # type: ignore[assignment]
                 self.hs.get_auth().get_user_by_access_token = get_user_by_access_token  # type: ignore[assignment]
                 self.hs.get_auth().get_access_token_from_request = Mock(  # type: ignore[assignment]
```
Comment (clokep) [inline tests/unittest.py:334]: I wonder if specifying `spec` would help ensure this is called the proper way...

Comment (clokep) [reply tests/unittest.py:334]: To expand on this a bit, I wonder if something like the following would be better at ensuring the callable looks like the "real" one:

```
patch.object(self.hs.get_auth(), "get_user_by_req",
    return_value=create_requester(
        UserID.from_string(self.helper.auth_user_id),
        token_id,
        False,
        False,
        None,
    ),
    autospec=True,
)
```

Comment (sandhose) [reply tests/unittest.py]: Just did that 👍

**Comment by clokep** [comment]
Ugh, something seems broken in the olddeps. 😢

**Comment by DMRobertson** [comment]
AsyncMock was added in 3.8; our oldest Python version is 3.7.

**Comment by sandhose** [comment]
> AsyncMock was added in 3.8; our oldest Python version is 3.7.

😢 

I tried something else: manually creating a future & returning that. I'm pretty sure I *have to* specify the Mock class, else `patch.object` will try to do an `AsyncMock` on Python 3.8+? 

Should I leave a note for when we deprecate support for Python 3.7?

**Comment by clokep** [comment]
We have a `make_awaitable` helper for this case. I think you can just do `return_value=make_awaitable(requester)` and it will be fine!

**Review by clokep — APPROVED** [review]

**Review by clokep — APPROVED** [review]
FTR this still looks good!

**Pull request merged**